hathor/cli/top.py

```diff
@@ -0,0 +1,704 @@
+        frac = total_time - floor(total_time)
+        seconds = floor(total_time)
+        hours = seconds // 3600
+        seconds = seconds % 3600
+        minutes = seconds // 60
+        seconds = seconds % 60
+
+        if hours >= 100:
+            return '{}h'.format(hours)
+        elif hours > 0:
+            return '{:02d}:{:02d}:{:02d}'.format(hours, minutes, seconds)
+        else:
+            return '{:02d}:{:02d}.{:02d}'.format(minutes, seconds, floor(100 * frac))
+
+
+def group_by_parent(proc_list: List[ProcItem]) -> ProcGroup:
+    """Group the processes by their parents.
+
+    It converts from the format received by the API and the format used by the printer.
+    """
+    d = defaultdict(list)
+    for proc in proc_list:
+        d[tuple(proc.key[:-1])].append(proc)
+    for k, v in d.items():
+        v.sort(key=lambda proc: proc.percent_cpu, reverse=True)
+    return d
+
+
+def group_cpu_percent(groups: ProcGroup, *, threshold: float = 0.5, separator: str = '!') -> ProcGroup:
+    """Group processes that consumes less than `threadhold` of CPU."""
+    new_groups: ProcGroup = defaultdict(list)
+    for key, children in groups.items():
+        hidden_procs: DefaultDict[str, List[ProcItem]] = defaultdict(list)
+        new_children: List[ProcItem] = []
+        for proc in children:
+            if proc.percent_cpu < threshold:
+                local_key = proc.key[-1]
+                parts = local_key.split(separator, 1)
+                if len(parts) == 1:
+                    gid = ''
+                else:
+                    gid = parts[0]
+                hidden_procs[gid].append(proc)
+            else:
+                new_children.append(proc)
+
+        for gid, proc_list in hidden_procs.items():
+            if not proc_list:
+                continue
+            if len(proc_list) == 1:
+                new_children += proc_list
+                continue
+            if gid:
+                local_key = '{}!({} hidden)'.format(gid, len(proc_list))
+            else:
+                local_key = '({} hidden)'.format(len(proc_list))
+
+            placeholder = ProcItem()
+            placeholder.key = tuple(key + (local_key,))
+            placeholder.percent_cpu = sum(x.percent_cpu for x in proc_list)
+            placeholder.total_time = sum(x.total_time for x in proc_list)
+            new_children.append(placeholder)
+        new_groups[key] = new_children
+    assert set(groups.keys()) == set(new_groups.keys())
+    return new_groups
+
+
+class Window(ABC):
+    """Abstract class for a window."""
+
+    def __init__(self, manager: 'ScreenManager', win: Any) -> None:
+        self.manager = manager
+        self.win = win
+
+    def on_data_update(self) -> None:
+        """Called when we get new data from the profiler."""
+        pass
+
+    def on_keypress(self, key: int) -> None:
+        """Called when the user press a key."""
+        pass
+
+    def render(self) -> None:
```

Comment:
nitpick: mark this as `@abstractmethod`, seems to make sense, otherwise a `Window()` can be created and `ABC` wouldn't even adding anything

Reply:
Great. Thanks! Done and also raised `NotImplementedError`.